rosbag2_storage/include/rosbag2_storage/qos.hpp

```diff
@@ -0,0 +1,164 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ROSBAG2_STORAGE__QOS_HPP_
+#define ROSBAG2_STORAGE__QOS_HPP_
+
+#include <map>
+#include <string>
+#include <vector>
+
+#include "rclcpp/node_interfaces/node_graph_interface.hpp"
+#include "rclcpp/qos.hpp"
+#include "yaml-cpp/yaml.h"
+
+
+#include "rosbag2_storage/visibility_control.hpp"
+
+namespace rosbag2_storage
+{
+/// Simple wrapper around rclcpp::QoS to provide a default constructor for YAML deserialization.
+class ROSBAG2_STORAGE_PUBLIC Rosbag2QoS : public rclcpp::QoS
+{
+public:
+  Rosbag2QoS()
+  : rclcpp::QoS(rmw_qos_profile_default.depth) {}
+
+  explicit Rosbag2QoS(const rclcpp::QoS & value)
+  : rclcpp::QoS(value) {}
+
+  Rosbag2QoS & default_history()
+  {
+    keep_last(rmw_qos_profile_default.depth);
+    return *this;
+  }
+
+  // Create an adaptive QoS profile to use for subscribing to a set of offers from publishers.
+  /**
+    * - Uses rosbag2_storage defaults for History since they do not affect compatibility.
+    * - Adapts Durability and Reliability, falling back to the least strict publisher when there
+    * is a mixed offer. This behavior errs on the side of forming connections with all publishers.
+    * - Does not specify Lifespan, Deadline, or Liveliness to be maximally compatible, because
+    * these policies do not affect message delivery.
+    */
+  static Rosbag2QoS adapt_request_to_offers(
+    const std::string & topic_name,
+    const std::vector<rclcpp::TopicEndpointInfo> & endpoints);
+
+  // Create a QoS profile to offer for playback.
+  /**
+    * This logic exists because rosbag2 does not record on a per-publisher basis, so we try to
+    * get as close as possible to the original system's behavior, given a single publisher.
+    * If all profiles are the same (excepting History & Lifespan, which are purely local),
+    * that exact value is returned.
+    * Otherwise, fall back to the rosbag2 default and emit a warning.
+    */
+  static Rosbag2QoS adapt_offer_to_recorded_offers(
+    const std::string & topic_name,
+    const std::vector<Rosbag2QoS> & profiles);
+};
+
+std::vector<rosbag2_storage::Rosbag2QoS> from_rclcpp_qos_vector(std::vector<rclcpp::QoS> in);
+std::string serialize_rclcpp_qos_vector(std::vector<rclcpp::QoS> in);
+std::vector<rclcpp::QoS> to_rclcpp_qos_vector(std::vector<rosbag2_storage::Rosbag2QoS> in);
+std::vector<rclcpp::QoS> to_rclcpp_qos_vector(YAML::Node in, int version);
+std::vector<rclcpp::QoS> to_rclcpp_qos_vector(std::string serielized, int version);
```

Comment:
```suggestion
std::vector<rosbag2_storage::Rosbag2QoS> 
from_rclcpp_qos_vector(const std::vector<rclcpp::QoS> & in);
std::string serialize_rclcpp_qos_vector(const std::vector<rclcpp::QoS> & in, int version = 9);
std::vector<rclcpp::QoS> to_rclcpp_qos_vector(const std::vector<rosbag2_storage::Rosbag2QoS> & in);
std::vector<rclcpp::QoS> to_rclcpp_qos_vector(const YAML::Node & in, int version);
std::vector<rclcpp::QoS> to_rclcpp_qos_vector(const std::string & serialized, int version);
```
Please use const references for parameters and fix typo in the `serialized` parameter